crates/python/src/api.rs

```diff
@@ -0,0 +1,223 @@
+use std::collections::HashMap;
+
+use pyo3::{
+    create_exception,
+    exceptions::{PyRuntimeError, PyValueError},
+    prelude::*,
+    pyfunction,
+    types::{PyComplex, PyDict, PyFloat, PyInt, PyList, PyString},
+    Py, PyResult,
+};
+use qcs::{
+    api::{
+        ExecutionResult, ExecutionResults, Register, RewriteArithmeticResult, TranslationResult,
+    },
+    qpu::{
+        quilc::{CompilerOpts, TargetDevice, DEFAULT_COMPILER_TIMEOUT},
+        Qcs,
+    },
+};
+use rigetti_pyo3::{
+    create_init_submodule, py_wrap_data_struct, py_wrap_error, py_wrap_type, py_wrap_union_enum,
+    wrap_error, ToPython,
+};
+
+create_init_submodule! {
+    classes: [
+        PyExecutionResult,
+        PyExecutionResults,
+        PyRegister,
+        PyRewriteArithmeticResult,
+        PyTranslationResult
+    ],
+    errors: [
+        InvalidConfigError,
+        ExecutionError,
+        TranslationError,
+        CompilationError,
+        RewriteArithmeticError,
+        DeviceIsaError,
+        QcsSubmitError
+    ],
+    funcs: [
+        compile,
+        rewrite_arithmetic,
+        translate,
+        submit,
+        retrieve_results,
+        build_patch_values,
+        get_quilc_version
+    ],
+}
+
+py_wrap_data_struct! {
+    PyRewriteArithmeticResult(RewriteArithmeticResult) as "RewriteArithmeticResult" {
+        program: String => Py<PyString>,
+        recalculation_table: Vec<String> => Py<PyList>
+    }
+}
+
+py_wrap_data_struct! {
+    PyTranslationResult(TranslationResult) as "TranslationResult" {
+        program: String => Py<PyString>,
+        ro_sources: Option<HashMap<String, String>> => Option<Py<PyDict>>
+    }
+}
+
+py_wrap_type! {
+    PyExecutionResults(ExecutionResults) as "ExecutionResults";
+}
+
+py_wrap_union_enum! {
+    PyRegister(Register) as "Register" {
+        f64: F64 => Vec<Py<PyFloat>>,
+        i16: I16 => Vec<Py<PyInt>>,
+        i32: I32 => Vec<Py<PyInt>>,
+        i8: I8 => Vec<Py<PyInt>>,
+        complex64: Complex64 => Vec<Py<PyComplex>>
```

Comment:
What's the reason to have this and [crates/python/src/register_data.rs#PyRegisterData](https://github.com/rigetti/qcs-sdk-rust/pull/230/files#diff-13066fc74c4b4c1cf4df7c4976361ec3a2ccdb990406e46b2b3fe27f7e2bb66eR8) and why do the `complex` types disagree?

Reply:
Having both is because the Rust library uses both. I am not sure why the variant names disagree.

Reply:
As you say, something that would need to be addressed upstream, so IMO resolved in terms of this MR.